### ISSUE TITLE
Search engine assignment: docker-compose build failed

### ISSUE THREAD

**Issue by Tonnpo** [mikeizbicki/cmc-csci143#184]
Hi, I encountered this issue when I ran
```
nohup docker-compose build &
```
I copied this directly from `nohup.out`
```
Installing collected packages: wcwidth, wasabi, tabulate, srsly, sortedcontainers, python-crfsuite, pymorphy2-dicts-uk, pymorphy2-dicts-ru, pymorphy2-dicts, py-cpuinfo, plac, murmurhash, jieba, docopt, DAWG-Python, cymem, chardet, certifi, zipp, urllib3, tqdm, tinydb, threadpoolctl, spacy-lookups-data, six, pyparsing, pymorphy2, pycparser, py, psutil, preshed, pluggy, numpy, more-itertools, joblib, idna, Cython, catalogue, attrs, sklearn-crfsuite, scipy, requests, packaging, importlib-metadata, dartsclone, cffi, blis, thinc, SudachiPy, scikit-learn, pythainlp, pytest, natto-py, SudachiDict-full, spacy, pyvi, pytest-benchmark
  Running setup.py install for cymem: started
  Running setup.py install for cymem: finished with status 'error'
^[[91m  error: subprocess-exited-with-error

  × Running setup.py install for cymem did not run successfully.
  │ exit code: 1
  ╰─> [11 lines of output]
      running install
      running build
      running build_py
      package init file 'cymem/tests/__init__.py' not found (or not a regular file)
      running build_ext
      building 'cymem.cymem' extension
      x86_64-linux-gnu-gcc -pthread -Wno-unused-result -Wsign-compare -DNDEBUG -g -fwrapv -O2 -Wall -g -ffile-prefix-map=/build/python3.9-RNBry6/python3.9-3.9.2=. -fstack-protector-strong -Wformat -Werror=format-security -g -fwrapv -O2 -g -ffile-prefix-map=/build/python3.9-RNBry6/python3.9-3.9.2=. -fstack-protector-strong -Wformat -Werror=format-security -Wdate-time -D_FORTIFY_SOURCE=2 -fPIC -I/usr/include/python3.9 -I/usr/include/python3.9 -c cymem/cymem.cpp -o build/temp.linux-x86_64-3.9/cymem/cymem.o -O3 -Wno-strict-prototypes -Wno-unused-function
      x86_64-linux-gnu-gcc: error: cymem/cymem.cpp: No such file or directory
      x86_64-linux-gnu-gcc: fatal error: no input files
      compilation terminated.
      error: command '/usr/bin/x86_64-linux-gnu-gcc' failed with exit code 1
      [end of output]

  note: This error originates from a subprocess, and is likely not a problem with pip.
^[[0m^[[91merror: legacy-install-failure

× Encountered error while trying to install package.
╰─> cymem

note: This is an issue with the package mentioned above, not pip.
hint: See above for output from the failure.
^[[0mThe command '/bin/sh -c cd /tmp  && git clone https://github.com/mikeizbicki/pspacy  && cd pspacy  && sh install_dependencies.sh  && make USE_PGXS=1  && make USE_PGXS=1 install' returned a non-zero code: 1
Service 'pg' failed to build : Build failed
```
Does anybody encounter the same issue or have a solution to this?

Thanks in advance!

**Comment by iamasteriix** [comment]
Same here. I'm not sure if it's related to the `FIXME`s in the pg service's `Dockerfile`, one of which is `pspacy`.

If you haven't edited it yet, the failing command from the error report starts at line 44 of `services/pg/Dockerfile`.

@Tonnpo I'd appreciate if you lmk if/when you find anything. I'm currently trying to add `libmecab-dev` that I found in [Mike's repo (`chajda`)](https://github.com/mikeizbicki/chajda), which is where https://github.com/mikeizbicki/pspacy redirects to. If not for that, I'm not sure we have proper documentation for how we're supposed to solve the issue.

**Comment by iamasteriix** [comment]
I ran `git clone https://github.com/mikeizbicki/pspacy  && cd pspacy  && sh install_dependencies.sh` locally and I think I found what the problem is.

@mikeizbicki I think there might be an issue with the `scipy` dependency version in the repository. I downgraded it to `1.5.4` and managed to (partly?) install and run the library (it asked for my password at the end for some reason and then threw the classic `This incident will be reported`).

**Comment by afroCoderHanane** [comment]
We spoke to Mike about it in office hours, he say he will look closely into it. Till then just relax and study for the final LOL

**Comment by mikeizbicki** [comment]
**tl;dr:** The problem is fixed, and recloning the repo should work

**long version:** We've talked about how the goal of Docker is to be able to have the same code run "anywhere"... unfortunately, it's much harder to have the same code run "anytime".  The `Dockerfile`s for this repo built fine for last year's version of this class.  But the `Dockerfile`s also depend on external services like debian apt repository's and python's pypi, and these repos often silently change the code that they will give you even when you ask for the same version.  Sometimes this is for security reasons (fixing critical exploits), but most often this is due to programming mistakes.  Maintaining consistent versioning is a hard problem that's basically ignored in cs education/research.  About 2 weeks ago, I went through and updated the `Dockerfile`s so that they would build again after the year's worth of bit rot... and everything worked when I published the assignment.  But alas something somewhere changed since I ran `git push` and you all ran `git clone`, causing the code to break once more.

This time, there was an issue in one of the dependencies that was updated sometime in the last few days.  I'm not 100% sure exactly which dependency broke, but the solution was to add the following line into the `Dockerfile`s to install some system libraries that were no longer getting installed:
```
RUN apt-get install -y  libxml2-dev libxslt1-dev
```
The problem was particularly nasty to debug because it was python's `pip install` command that was erroring, but the error had nothing to do with python AFAICT and had to do with settings in the upstream apt repos causing these libraries to not get installed anymore.

**Comment by Tonnpo** [comment]
Hi @mikeizbicki ,

I recloned the repo and follow the setup instructions. I am not sure if this is specific to my bigdata folder. When I ran
```
nohup docker-compose build &
```
It seems that it reaches this step before getting the error
```
Step 11/16 : RUN cd /tmp  && git clone --filter=blob:none --no-checkout https://github.com/mikeizbicki/metahtml  && cd metahtml  && git sparse-checkout init  && git sparse-checkout set '!/tests/.cache' '/*'  && git checkout master  && pip3 install -r requirements.txt  && python3 scripts/generate_tests_json.py > /tmp/metahtml/all_tests.json
```
And here is the error from `nohup.out`
```
Collecting cdx-toolkit==0.9.29
  Downloading cdx_toolkit-0.9.29.tar.gz (33 kB)
  Preparing metadata (setup.py): started
  Preparing metadata (setup.py): finished with status 'done'
Collecting certifi==2020.11.8
      cc1plus: warning: command-line option ‘-Wno-strict-prototypes’ is valid for C/ObjC but not for C++
      x86_64-linux-gnu-g++ -pthread -shared -Wl,-O1 -Wl,-Bsymbolic-functions -Wl,-z,relro -g -fwrapv -O2 build/temp.linux-x86_64-cpython-39/spacy/tokens/morphanalysis.o -L/usr/lib -o build/lib.linux-x86_64-cpython-39/spacy/tokens/morphanalysis.cpython-39-x86_64-linux-gnu.so
      building 'spacy.tokens._retokenize' extension
      x86_64-linux-gnu-gcc -pthread -Wno-unused-result -Wsign-compare -DNDEBUG -g -fwrapv -O2 -Wall -g -ffile-prefix-map=/build/python3.9-RNBry6/python3.9-3.9.2=. -fstack-protector-strong -Wformat -Werror=format-security -g -fwrapv -O2 -fPIC -I/usr/include/python3.9 -I/tmp/pip-install-njohjlh6/spacy_40d2adc22ad24408ba19782812e738db/include -I/usr/include/python3.9 -c spacy/tokens/_retokenize.cpp -o build/temp.linux-x86_64-cpython-39/spacy/tokens/_retokenize.o -O2 -Wno-strict-prototypes -Wno-unused-function
      cc1plus: warning: command-line option ‘-Wno-strict-prototypes’ is valid for C/ObjC but not for C++
      In file included from /tmp/pip-install-njohjlh6/spacy_40d2adc22ad24408ba19782812e738db/include/numpy/ndarraytypes.h:1728,
                       from /tmp/pip-install-njohjlh6/spacy_40d2adc22ad24408ba19782812e738db/include/numpy/ndarrayobject.h:17,
                       from /tmp/pip-install-njohjlh6/spacy_40d2adc22ad24408ba19782812e738db/include/numpy/arrayobject.h:15,
                       from spacy/tokens/_retokenize.cpp:702:
      /tmp/pip-install-njohjlh6/spacy_40d2adc22ad24408ba19782812e738db/include/numpy/npy_deprecated_api.h:11:2: warning: #warning "Using deprecated NumPy API, disable it by #defining NPY_NO_DEPRECATED_API NPY_1_7_API_VERSION" [-Wcpp]
         11 | #warning "Using deprecated NumPy API, disable it by #defining NPY_NO_DEPRECATED_API NPY_1_7_API_VERSION"
            |  ^~~~~~~
      spacy/tokens/_retokenize.cpp:37085:1: fatal error: error writing to /tmp/ccI5k5yl.s: Disk quota exceeded
      37085 | }
            | ^
      compilation terminated.
      error: command '/usr/bin/x86_64-linux-gnu-gcc' failed with exit code 1
      [end of output]

  note: This error originates from a subprocess, and is likely not a problem with pip.
  ERROR: Failed building wheel for spacy
Building wheel for py-cpuinfo (setup.py): finished with status 'done'
  Created wheel for py-cpuinfo: filename=py_cpuinfo-8.0.0-py3-none-any.whl size=22245 sha256=9004e071ec7ff5d73dd6da11a24c4ba7baeec0c9b0946b82e5b01b0934c42e82
  Stored in directory: /root/.cache/pip/wheels/a9/33/c2/bcf6550ff9c95f699d7b2f261c8520b42b7f7c33b6e6920e29
Successfully built cffi dartsclone docopt jieba murmurhash natto-py numpy preshed psutil python-crfsuite scikit-learn spacy-lookups-data srsly SudachiPy wasabi pymorphy2 py-cpuinfo
Failed to build SudachiDict-full spacy
ERROR: Could not build wheels for spacy, which is required to install pyproject.toml-based projects
The command '/bin/sh -c cd /tmp  && git clone https://github.com/mikeizbicki/pspacy  && cd pspacy  && sh install_dependencies.sh  && make USE_PGXS=1  && make USE_PGXS=1 install' returned a non-zero code: 1
Service 'pg' failed to build : Build failed
```

Has anyone encountered the same issue?

Thanks very much!
